### ISSUE TITLE
multithreaded http2 GET occasionally throws IllegalStateException

### ISSUE THREAD

**Issue by ifedorenko** [square/okhttp#3521]
I get infrequent exceptions like below when downloading artifacts from nginx on multiple threads over http2. The problem is likely related to (or at least made more severe by) nginx sending GOAWAY frames to force the client acquire new connection. I am using okhttp 3.9.0-SNAPSHOT with swankjesse's fix for #3516. I can provide nginx config and a small java app that triggers the problem semi-reliably but don't have proper standalone test.

```
java.lang.IllegalStateException
	at okhttp3.internal.connection.StreamAllocation.acquire(StreamAllocation.java:419)
	at okhttp3.ConnectionPool.get(ConnectionPool.java:126)
	at okhttp3.OkHttpClient$1.get(OkHttpClient.java:152)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:175)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:133)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:112)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:202)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at test.okhttp.DownloadTest.download(DownloadTest.java:96)
	at test.okhttp.DownloadTest.lambda$0(DownloadTest.java:77)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
```

All download threads share the same `OkHttpClient` instance and each download thread runs this (abbreviated) code:

```java
Request request = new Request.Builder() //
    .url("https://localhost:8443/...") //
    .build();
try (Response response = client.newCall(request).execute()) {
    if (response.isSuccessful()) {
        response.body().string();
    }
}    
```

I also see another exception that may or may not be related:

```
java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:8443
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:240)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:158)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:239)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:133)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:112)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:202)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at test.okhttp.DownloadTest.download(DownloadTest.java:96)
	at test.okhttp.DownloadTest.lambda$0(DownloadTest.java:77)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:125)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:238)
	... 23 more
```

**Comment by dave-r12** [comment]
@ifedorenko do any of the requests result in a redirect or auth challenge?

**Comment by ifedorenko** [comment]
No, no redirects and no auth challenges according to nginx logs (all requests have `status=200`).

**Comment by dave-r12** [comment]
Gotcha -- if you want some help feel free to provide the config and test case. I can help take a look.

**Comment by ifedorenko** [comment]
I ran more tests and see rather wide variety of exceptions thrown when nginx tells okhttp GOWAY. I believe specific exception depends on state of individual http2 stream and timing of GOAWAY.

`IllegalStateException` is fixed by #3525 and I believe happens when one failed stream marks the connection as noNewStreams while another failed stream attempts recovery using the same connection.

All other exceptions are subclasses of `IOException` and while annoying do not cause real problems because they are already covered by retry-with-backoff logic in my code. I can document the exceptions I see in my test if anyone is interested.

**Comment by swankjesse** [comment]
Nice!
